### DISCUSSION TITLE
oidc dyncamic client registration

### DISCUSSION THREAD

**Discussion by cyberb** [authelia/authelia#7304] [General]
I am working on a personal self-hosting platform https://github.com/syncloud/platform where you can install various open source apps like nextcloud for personal use.
I am trying to add authelia as a single sign on, everything looks great and one of the first apps using it will be peertube.
As I have many apps in the store I would like to know what is the best approach right now to register each app/client in a more automated/dynamic/safe way?

1. I do not want platform/auth to know all apps upfront in its yaml.
2. Adding client dynamically to the authelia yaml when user installs the app is an option, but a bit dangerous as one wrong app registration will probably break all installed apps and ideally I want apps to be isolated (with their problems).
3. I heard sql is an option for authelia config, is there any validation of each client registration or will one wrong client config affect the whole authelia?

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To address your requirements for dynamically registering each app/client in a safe and automated way with Authelia, here are some considerations and solutions based on the information available:

1. **Avoiding Static Configuration in YAML**: You mentioned that you do not want the platform/auth to know all apps upfront in its YAML configuration. This is aligned with the principle of dynamic client registration, which Authelia supports to some extent. However, Authelia traditionally relies on a static configuration file (YAML) for defining clients.

2. **Dynamic Client Registration**: While Authelia's documentation does not explicitly mention support for OpenID Connect Dynamic Client Registration, which would allow clients to register themselves without predefined configurations, you can simulate a dynamic-like behavior using a database.

3. **Using a Database for Dynamic Configuration**:
   - **SQL Database**: Authelia supports using SQL databases like MySQL, MariaDB, or PostgreSQL for storing session and other configurations. This can be leveraged to manage client configurations more dynamically.
   - **Client Registration Validation**: When using a database, each client registration can be isolated in its own transaction, reducing the risk that a misconfiguration in one client affects others. However, it's crucial to implement proper validation logic either in the application logic that interacts with the database or within stored procedures in the database.

4. **Security and Isolation**: To ensure that the registration of new clients does not impact the stability or security of the platform, consider implementing additional layers of validation and testing before a new client configuration is committed to the production database. This could include automated tests or manual review processes.

5. **Automation and Safety**: To automate the process safely, you could develop a middleware component that handles the registration requests from new apps. This component would validate the registration details against a set of predefined rules before updating the database. This approach helps in isolating the impact of any erroneous registrations.

Given these points, the best approach right now would be to use a SQL database for storing client configurations and to develop a robust middleware system that can handle and validate client registrations dynamically. This setup allows for scalability and reduces the risks associated with manual updates to configuration files.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Dynamic Client Registration not supported at the moment but is a planned feature and will be supported via the specification prescribed method noted in [RFC7591: OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591) and [OpenID Connect Dynamic Client Registration 1.0](https://openid.net/specs/openid-connect-registration-1_0.html), specifically via a OAuth 2.0 Flow which grants a token with specific characteristics (audience and scope).

This will require an in-database client table, which will have to exclusively be used to properly support it.

**Reply by riedel** [reply]
Would it make sense to add an actual feature request as issue? Just saying, I also ran into a use case for which manual registration with the absence of wildcards becomes probably infeasible (providing an intranet template VM with authentication preconfigured, I am currently trying to use templates for this).

**Comment by cyberb** [comment]
Thanks, are there any estimates, is it a month or a year?

**Reply by flixman** [reply]
@cyberb: why is this discussion closed? I have managed to get this behavior in the poor man's way: A script that puts together the fragments and restarts gitea when the folder where the fragments are changes. It is far from perfect, but kind of works. I was considering doing this with a database, so that multiple clients can be updated concurrently and the writes to the config and the restart of authelia get consolidated. Where you able to do this in another way?

@james-d-elliott Do you have any estimate for the implementation of those activities you explained before? Can we help in any way?

**Reply by cyberb** [reply]
I did the same in the end, maintain a db of clients and regenerate config every time it changes.

https://github.com/syncloud/platform/blob/4b1d679c59cbb52b361ea61d216922a8299fa28a/backend/auth/authelia.go#L126

**Reply by james-d-elliott** [reply]
I have [written](https://github.com/authelia/oauth2-provider/tree/master/handler/rfc7591) most of the spec out already. Chances are it'll land soon.

**Reply by flixman** [reply]
@james-d-elliott do you need any help with that?
